numba/annotations/type_annotations.py

```diff
@@ -0,0 +1,271 @@
+from __future__ import print_function, absolute_import
+import inspect
+import re
+from collections import Mapping, defaultdict
+import textwrap
+from contextlib import closing
+from numba.io_support import StringIO
+from numba import ir
+import numba.dispatcher
+import os
+import sys
+
+
+class SourceLines(Mapping):
+    def __init__(self, func):
+
+        try:
+            lines, startno = inspect.getsourcelines(func)
+        except IOError:
+            self.lines = ()
+            self.startno = 0
+        else:
+            self.lines = textwrap.dedent(''.join(lines)).splitlines()
```

Comment:
This mapping class seems to have precomputed read-only contents, so why not create a standard dictionary instead?
(or a defaultdict if you want to return '' on unknown line numbers)